gpu/octree/src/cuda/approx_nsearch.cu

```diff
@@ -105,176 +113,96 @@ namespace pcl { namespace device { namespace appnearest_search
-		{                        						
-            __shared__ volatile float dist2[CTA_SIZE];
-            __shared__ volatile int   index[CTA_SIZE];
-			
-            int tid = threadIdx.x;
-			dist2[tid] = std::numeric_limits<float>::max();
-
-			//serial step
-            for (int idx = Warp::laneId(); idx < length; idx += Warp::STRIDE)
-			{
-				float dx = points[idx                  ] - active_query.x;
-				float dy = points[idx + points_step    ] - active_query.y;
-				float dz = points[idx + points_step * 2] - active_query.z;
-
-				float d2 = dx * dx + dy * dy + dz * dz;
-
-				if (dist2[tid] > d2)
-				{
-					dist2[tid] = d2;
-					index[tid] = idx;                            
-				}
-			}
-			//parallel step
-			unsigned int lane = Warp::laneId();
-
-			float mind2 = dist2[tid];
-
-			if (lane < 16)
-			{
-				float next = dist2[tid + 16];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 16]; 
-				}                        
-			}
-
-			if (lane < 8)
-			{
-				float next = dist2[tid + 8];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 8]; 
-				}                        
-			}
-
-			if (lane < 4)
-			{
-				float next = dist2[tid + 4];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 4]; 
-				}                        
-			}
-
-			if (lane < 2)
-			{
-				float next = dist2[tid + 2];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 2]; 
-				}                        
-			}
-
-			if (lane < 1)
-			{
-				float next = dist2[tid + 1];
-				if (mind2 > next) 
-				{ 
-					dist2[tid] = mind2 = next; 
-					index[tid] = index[tid + 1]; 
-				}                        
-			}        
-
-			return index[tid - lane];
-		}
+        template <int CTA_SIZE>
+        __device__  std::pair<int, float>
+        NearestWarpKernel(const int beg,
+                          const int field_step,
+                          const int length,
+                          const float3& active_query)
+
+        {
+          int index = 0;
+          float dist2 = std::numeric_limits<float>::max();
+
+          // serial step
+          for (int idx = Warp::laneId(); idx < length; idx += Warp::STRIDE) {
+            const float dx = batch.points[beg + idx] - active_query.x;
+            const float dy = batch.points[beg + idx + field_step] - active_query.y;
+            const float dz = batch.points[beg + idx + field_step * 2] - active_query.z;
+
+            const float d2 = dx * dx + dy * dy + dz * dz;
+
+            if (dist2 > d2) {
+              dist2 = d2;
+              index = idx;
+            }
+          }
+
+          // find minimum distance among warp threads
+          constexpr unsigned FULL_MASK = 0xFFFFFFFF;
+          static_assert(KernelPolicy::WARP_SIZE <= 8*sizeof(unsigned int));
```

Comment:
```suggestion
          static_assert(KernelPolicy::WARP_SIZE <= 8*sizeof(FULL_MASK));
```

`static_assert` without a message is a C++17 feature. You need to add a message in case this assert fails.